client/matching/loadbalancer.go

```diff
@@ -135,6 +140,130 @@ func (lb *defaultLoadBalancer) pickPartition(
 		return taskQueue.GetName()
 	}
 
-	n := max(1, nPartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
+	n := max(1, lb.nWritePartitions(nsName.String(), tqName.BaseNameString(), taskQueueType))
 	return tqName.WithPartition(rand.Intn(n)).FullName()
 }
+
+// PickReadPartition picks a partition for poller to poll task from, and keeps load balanced between partitions.
+// Caller is responsible to call pollToken.Release() after complete the poll.
+func (lb *defaultLoadBalancer) PickReadPartition(
+	namespaceID namespace.ID,
+	taskQueue taskqueuepb.TaskQueue,
+	taskQueueType enumspb.TaskQueueType,
+	forwardedFrom string,
+) *pollToken {
+	if forwardedFrom != "" || taskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
+		// no partition for sticky task queue and forwarded request
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	parsedName, err := tqname.Parse(taskQueue.GetName())
+	if err != nil || err == nil && !parsedName.IsRoot() {
+		// parse error or partition already picked, use as-is
+		return &pollToken{fullName: taskQueue.GetName()}
+	}
+
+	key := taskQueueKey{NamespaceID: namespaceID, Name: parsedName, Type: taskQueueType}
+
+	lb.lock.Lock()
+	tqlb, ok := lb.taskQueueLBs[key]
+	if !ok {
+		tqlb = newTaskQueueLoadBalancer(key)
+		lb.taskQueueLBs[key] = tqlb
+	}
+	lb.lock.Unlock()
+
+	// For read path it's safer to return global default partition count instead of root partition, when we fail to
+	// map namespace ID to name.
+	var partitionCount = dynamicconfig.GlobalDefaultNumTaskQueuePartitions
+
+	namespaceName, err := lb.namespaceIDToName(namespaceID)
+	if err == nil {
+		partitionCount = lb.nReadPartitions(string(namespaceName), parsedName.BaseNameString(), taskQueueType)
+	}
+
+	return tqlb.pickReadPartition(partitionCount, lb.forceReadPartition())
+}
+
+func newTaskQueueLoadBalancer(key taskQueueKey) *tqLoadBalancer {
+	return &tqLoadBalancer{
+		taskQueue: key,
+		lock:      sync.Mutex{},
+	}
+}
+
+func (b *tqLoadBalancer) pickReadPartition(partitionCount int, forcedPartition int) *pollToken {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	// ensure we reflect dynamic config change if it ever happens
+	b.ensurePartitionCountLocked(partitionCount)
+
+	partitionID := forcedPartition
+
+	if partitionID < 0 {
+		partitionID = b.pickReadPartitionWithFewestPolls(partitionCount)
+	}
+
+	b.pollerCounts[partitionID]++
+
+	return &pollToken{
+		fullName:    b.taskQueue.Name.WithPartition(partitionID).FullName(),
+		partitionID: partitionID,
+		balancer:    b,
+	}
+}
+
+// caller to ensure that lock is obtained before call this function
+func (b *tqLoadBalancer) pickReadPartitionWithFewestPolls(partitionCount int) int {
```

Comment:
How many partitions can we have? If the answer is "a lot" we may want to look at a randomized approach; something like https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/ ([original paper](https://www.eecs.harvard.edu/~michaelm/postscripts/tpds2001.pdf))

Reply:
Thanks @tdeebswihart, I learned this today. I feel what we have is ok, at least for now. almost all use cases should be fine with a few partitions. I don't imagine anyone needs more than a 100.

Reply:
Do we have data from a production cluster for how many task queue partitions we have? Say, just the max partitions across all task queues

Reply:
Yimin replied with `256` in Slack